text/0000-rust-analyzer.md

```diff
@@ -0,0 +1,155 @@
+
+However, there are a number of practical concerns with taking that approach. One concern is that, for rust-analyzer's current users, it would represent a regression. rust-analyzer would no longer be available (or at least no longer updated) and it would take some time until rustc is at feature parity. Moreover, experience has shown that refactoring rustc can move relatively slowly, simply due to the age of the codebase, the amount of code involved, and the complicated non-standard build process.
+
+Further, the 'reimplement' approach would represent a "constraint" on the ordering in which we do our work. With the design proposed in this RFC, for example, rust-analyzer is able to make use of the chalk library already. This is only possible because rust-analyzer has a "stub" version of Rust's name resolution engine and type checker embedded in it -- this type checker is not perfect, but it's good enough to drive chalk and gain useful experience. This allows us to create an end-to-end IDE user experience sooner, in effect.
+
+In contrast, if we were to try and rebuild rust-analyzer within rustc, even if we had rustc adopt chalk or some other IDE-friendly trait resolution algorithm, that would not be of use to IDE users until we had also upgraded the name resolution algorithm and type checker to be IDE friendly. In short, having a "prototype" version of these algorithms that lives in rust-anaylzer is both a pro and a con: it means we have to maintain two versions, but it means users get benefits faster and developers can experiment more freely.
+
+# Prior art
+[prior-art]: #prior-art
+
+The current proposal is informed by experience with existing RLS and query-based compilation in rustc. Additionally, rust-analyzer heavily draws from lessons learned while developing IntelliJ Rust.
+
+It's interesting that many compilers went through a phase with parallel implementaitons to get a great IDE support
+
+* For C#, the [Roslyn](https://github.com/dotnet/roslyn) project was a from scratch implemenation.
+* [Dart for a long time had different front-ends for command line and interractive compilers](https://youtu.be/WjdrUphF5l4?t=2204)
+* [Swift is transitioning to new syntax tree library by "reimplement separately, then swap" approach](https://medium.com/@kitasuke/deep-dive-into-integrating-libsyntax-into-the-compiler-pipeline-2d478c8600a1)
+
+Notable exceptions:
+
+* Kotlin, TypeScript -- these languages were implemented with IDEs in mind from the start
+* OCaml with merlin and C++ with clangd -- languages with header files and forward declarations make it easier to adapt traditional compiler architecture to certain IDE tasks like completion.
```

Comment:
Note, that the OCaml had to rethink it's merlin and LSP architecture, thus the [ocaml-lsp](https://github.com/ocaml/ocaml-lsp) project. As it turned out, the original Merlin architecture wasn't suitable enough for the VSCode and Vim/Emacs LSP clients.